consensus/src/threshold_simplex/actors/voter/actor.rs

```diff
@@ -15,273 +13,201 @@ use commonware_cryptography::{
     bls12381::primitives::{
         group::{self, Element},
         ops::{threshold_signature_recover, threshold_signature_recover_pair},
-        poly,
         variant::Variant,
     },
     Digest, Scheme,
 };
 use commonware_macros::select;
 use commonware_p2p::{
     utils::codec::{wrap, WrappedSender},
-    Receiver, Recipients, Sender,
+    Blocker, Receiver, Recipients, Sender,
+};
+use commonware_runtime::{
+    telemetry::metrics::histogram::{self, Buckets},
+    Clock, Handle, Metrics, Spawner, Storage,
 };
-use commonware_runtime::{Clock, Handle, Metrics, Spawner, Storage};
 use commonware_storage::journal::variable::{Config as JConfig, Journal};
 use commonware_utils::quorum;
+use core::panic;
 use futures::{
     channel::{mpsc, oneshot},
     future::Either,
     pin_mut, StreamExt,
 };
-use prometheus_client::metrics::{
-    counter::Counter, family::Family, gauge::Gauge, histogram::Histogram,
-};
+use prometheus_client::metrics::{counter::Counter, gauge::Gauge, histogram::Histogram};
 use rand::Rng;
-use std::{collections::hash_map::Entry, sync::atomic::AtomicI64};
+use std::sync::{atomic::AtomicI64, Arc};
 use std::{
-    collections::{BTreeMap, HashMap},
+    collections::BTreeMap,
     time::{Duration, SystemTime},
 };
 use tracing::{debug, trace, warn};
 
 const GENESIS_VIEW: View = 0;
 
+/// Compute the quorum for a given polynomial.
+fn polynomial_quorum<P>(polynomial: &[P]) -> u32 {
+    let n = polynomial.len() as u32;
+    quorum(n)
+}
+
+/// Action to take after processing a message.
+enum Action {
+    /// Skip processing the message.
+    Skip,
+    /// Block the peer from sending any more messages.
+    Block,
+    /// Process the message.
+    Process,
+}
+
 struct Round<
+    E: Clock,
     C: Scheme,
     V: Variant,
     D: Digest,
-    R: Reporter<Activity = Activity<V, D>>,
     S: ThresholdSupervisor<
         Seed = V::Signature,
         Index = View,
         Share = group::Share,
         PublicKey = C::PublicKey,
+        Identity = V::Public,
     >,
 > {
     start: SystemTime,
-    reporter: R,
     supervisor: S,
 
     view: View,
-    participants: usize,
+    quorum: u32,
+
+    // Leader is set as soon as we know the seed for the view.
+    leader: Option<(C::PublicKey, u32)>,
+
+    requested_proposal_build: bool,
+    requested_proposal_verify: bool,
+    verified_proposal: bool,
+
+    proposal: Option<Proposal<D>>,
 
-    leader: Option<C::PublicKey>,
     leader_deadline: Option<SystemTime>,
     advance_deadline: Option<SystemTime>,
     nullify_retry: Option<SystemTime>,
 
-    // Track one proposal per view (only matters prior to notarization)
-    proposal: Option<Proposal<D>>,
-    requested_proposal: bool,
-    verified_proposal: bool,
-
-    // Track notarizes for all proposals (ensuring any participant only has one recorded notarize)
-    notarized_proposals: HashMap<Proposal<D>, Vec<u32>>,
```

Comment:
Because we only learn about messages once verified, this logic is now MUCH simpler.